lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+                  limit=None, is_regex=False):
+    """Build the SELECT query to be executed"""
+    args = []
+
+    fields = ""
+    criteria = "WHERE"
+    sorting = ""
+    tables = "FROM aff4"
+
+    subject = utils.SmartUnicode(subject)
+
+    #Set fields, tables, and criteria and append args
+    if is_regex:
+      tables += " JOIN attributes ON aff4.attribute_hash=attributes.hash"
+      regex = re.match(r'(^[a-zA-Z0-9_]+:[a-zA-Z0-9_:]*)(.*)', predicate)
+      if regex:
+        #If predicate has prefix then break down for query optimizations
+        parts = predicate.split(":", 1)
+
+        #Add prefix usage for all regex queries
+        criteria += " aff4.subject_hash=unhex(md5(%s)) AND aff4.prefix=(%s)"
+        args.append(subject)
+        args.append(parts[0])
+        #If the remainder after prefix is not a match all regex then break it
+        #down into like and rlike components
+        if not (parts[1] == ".*" or parts[1] == ".+"):
+          like = regex.groups()[0] + "%"
+          rlike = regex.groups()[1]
+          #Add like if available
+          if like:
+            criteria += " AND attributes.attribute like %s"
+            args.append(like)
+          #add rlike if available
+          if rlike:
+            criteria += " AND attributes.attribute rlike %s"
+            args.append(rlike)
+      else:
+        #If predicate has no prefix just rlike
+        criteria += " aff4.subject_hash=unhex(md5(%s)) AND attributes.attribute rlike %s"
+        args.append(subject)
+        args.append(predicate)
+    else:
+      if predicate:
+        criteria += " aff4.subject_hash=unhex(md5(%s)) AND aff4.attribute_hash=unhex(md5(%s))"
+        args.append(subject)
+        args.append(predicate)
+      else:
+        criteria += " aff4.subject_hash=unhex(md5(%s))"
+        args.append(subject)
+
+    #Limit to time range if specified
+    if isinstance(timestamp, (tuple, list)):
+      criteria += " AND aff4.timestamp >= %s AND aff4.timestamp <= %s"
+      args.append(int(timestamp[0]))
+      args.append(int(timestamp[1]))
+
+    if action == "SELECT":
+      fields = "aff4.value, aff4.timestamp"
+      if is_regex:
+        fields += ", attributes.attribute"
+
+      #Modify fields and sorting for timestamps
+      if timestamp is None or timestamp == self.NEWEST_TIMESTAMP:
+        tables += " JOIN (SELECT attribute_hash, MAX(timestamp) timestamp \
+" + tables + " " + criteria + " GROUP BY attribute_hash) maxtime on \
+aff4.attribute_hash=maxtime.attribute_hash and \
+aff4.timestamp=maxtime.timestamp"
+        criteria = "WHERE aff4.subject_hash=unhex(md5(%s))"
+        args.append(subject)
+      else:
+        #Always order results
+        sorting = "ORDER BY aff4.timestamp DESC"
+      #Add limit if set
+      if limit:
+        sorting += " LIMIT %s" % int(limit)
+    else:
+      fields = "aff4"
+
+    query = action + " " + fields + " " + tables + " " + criteria + " " + sorting
```

Comment:
you could use " ".join([action, fields, tables, criteria, sorting]). Also, I feel like string interpolation is nicer than using + concatenation for most cases. Would you mind changing to that?